File: apps/android/VLinkerOBD/app/src/main/java/com/jslee1972/vlinkerobd/obd/ObdCommandQueue.kt
package com.jslee1972.vlinkerobd.obd

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout

/**
 * Serializes ELM/STN command execution: only one command is ever in flight, each attempt waits
 * for the `>` prompt (or [ObdCommand.timeoutMs]), and a timed-out attempt is retried once before
 * the command is reported as [ObdCommandResult.Timeout]. A manual command is queued the same as
 * any other — callers (the ViewModel) decide whether to pause automatic polling first.
 *
 * Incoming bytes are forwarded into [incomingChannel] by one persistent collector (started once,
 * outliving any single attempt) rather than each attempt subscribing to [ObdTransport.incoming]
 * directly. That persistence matters on slow, non-CAN links (ISO9141/KWP-era vehicles observed in
 * the field): a response can arrive just after its attempt gave up to a timeout/retry, and a
 * per-attempt `Flow.first{}` subscription has no way to catch bytes that land in that gap — they
 * either vanish or, worse, get picked up by the *next* attempt's fresh subscription and get
 * concatenated onto an unrelated response (surfaced as "解析失敗"/garbled multi-response logs).
 * Routing everything through one durable channel and explicitly draining it with
 * [drainStaleIncoming] before every write (first attempt or retry) guarantees a command always
 * starts reading from a clean slate.
 */
class ObdCommandQueue(
    private val transport: ObdTransport,
    private val scope: CoroutineScope,
) {
    private data class PendingCommand(
        val command: ObdCommand,
        val result: CompletableDeferred<ObdCommandResult>,
    )

    private sealed interface AttemptOutcome {
        data class Prompted(val raw: String) : AttemptOutcome
        data object TimedOut : AttemptOutcome
        data object Cancelled : AttemptOutcome
    }

    private val channel = Channel<PendingCommand>(Channel.UNLIMITED)
    private val incomingChannel = Channel<ByteArray>(Channel.UNLIMITED)
    private var currentAttempt: Deferred<*>? = null

    init {
        scope.launch {
            transport.incoming.collect { chunk -> incomingChannel.send(chunk) }
        }
        scope.launch {
            for (pending in channel) {
                pending.result.complete(runWithRetry(pending.command))
            }
        }
    }

    suspend fun execute(command: ObdCommand): ObdCommandResult {
        val deferred = CompletableDeferred<ObdCommandResult>()
        channel.send(PendingCommand(command, deferred))
        return deferred.await()
    }

    /** Cancels the attempt currently in flight (if any); the command resolves as [ObdCommandResult.Cancelled]. */
    fun cancelCurrent() {
        currentAttempt?.cancel()
    }

    fun close() {
        channel.close()
        incomingChannel.close()
        currentAttempt?.cancel()
    }

    /** Discards whatever is already sitting in [incomingChannel] — the tail of a previous attempt's
     * response that arrived too late to be read by it. Non-suspending, so it never delays a fresh
     * write and never risks eating bytes that belong to the command about to be sent. */
    private fun drainStaleIncoming() {
        while (incomingChannel.tryReceive().isSuccess) {
            // Deliberately discarded — see class doc for why this must happen before every write.
        }
    }

    private suspend fun runWithRetry(command: ObdCommand): ObdCommandResult {
        repeat(command.maxAttempts) { attemptIndex ->
            when (val outcome = runSingleAttempt(command)) {
                is AttemptOutcome.Prompted -> return ObdCommandResult.Success(outcome.raw)
                AttemptOutcome.Cancelled -> return ObdCommandResult.Cancelled
                AttemptOutcome.TimedOut -> if (attemptIndex == command.maxAttempts - 1) {
                    return ObdCommandResult.Timeout
                }
            }
        }
        return ObdCommandResult.Timeout
    }

    private suspend fun runSingleAttempt(command: ObdCommand): AttemptOutcome {
        drainStaleIncoming()
        val bytes = (command.text.trim().uppercase() + "\r").toByteArray()
        transport.write(bytes)

        val attempt = scope.async {
            withTimeout(command.timeoutMs) {
                val buffer = StringBuilder()
                while (!buffer.contains(">")) {
                    buffer.append(String(incomingChannel.receive()))
                }
                buffer.toString()
            }
        }
        currentAttempt = attempt

        return try {
            AttemptOutcome.Prompted(attempt.await())
        } catch (timeout: TimeoutCancellationException) {
            AttemptOutcome.TimedOut
        } catch (cancellation: CancellationException) {
            AttemptOutcome.Cancelled
        } finally {
            currentAttempt = null
        }
    }
}

File: apps/ios/VLinkerOBD/ObdCommandQueue.swift
import Foundation

/// A byte-stream transport an `ObdCommandQueue` can write to and receive decoded text chunks
/// from. `OBDBLEManager` implements this over CoreBluetooth.
protocol ObdTransport: AnyObject {
    /// Set once by the queue; called with each decoded text chunk as it arrives.
    var onIncoming: ((String) -> Void)? { get set }
    func write(_ data: Data)
}

struct ObdCommand {
    var text: String
    var timeoutMs: Int = 2000
    var maxAttempts: Int = 2
}

enum ObdCommandResult: Equatable {
    case success(String)
    case timeout
}

/// Serializes ELM/STN command execution: only one command is ever in flight, each attempt waits
/// for the `>` prompt (or `timeoutMs`), and a timed-out attempt is retried once before the
/// command is reported as `.timeout`. Mirrors the guarantees of Android's `ObdCommandQueue.kt`
/// (single Channel-based FIFO on a main-immediate dispatcher) using a `@MainActor`-isolated class
/// instead — everything here runs cooperatively on the main actor, so there's no true data race
/// to guard against, only ordering to preserve.
///
/// Incoming bytes are appended to a durable buffer (`incomingBuffer`) rather than read directly
/// off the transport by each attempt, and that buffer is explicitly cleared (`drainStaleIncoming`)
/// before every write — the same fix Android needed for slow, non-CAN links where a response can
/// arrive just after its attempt already timed out: without draining, that stale response would
/// get concatenated onto the next command's reply.
@MainActor
final class ObdCommandQueue {
    private let transport: ObdTransport
    private var incomingBuffer = ""
    /// Fires (once) when new data arrives while an attempt is waiting — see `waitForIncoming`.
    private var wakeUp: (() -> Void)?

    private var commandQueue: [(ObdCommand, CheckedContinuation<ObdCommandResult, Never>)] = []
    private var isProcessing = false

    init(transport: ObdTransport) {
        self.transport = transport
        transport.onIncoming = { [weak self] text in
            Task { @MainActor in self?.receiveIncoming(text) }
        }
    }

    func execute(_ command: ObdCommand) async -> ObdCommandResult {
        await withCheckedContinuation { continuation in
            commandQueue.append((command, continuation))
            processNextIfNeeded()
        }
    }

    private func receiveIncoming(_ text: String) {
        incomingBuffer += text
        let wake = wakeUp
        wakeUp = nil
        wake?()
    }

    private func drainStaleIncoming() {
        incomingBuffer = ""
    }

    private func processNextIfNeeded() {
        guard !isProcessing, !commandQueue.isEmpty else { return }
        isProcessing = true
        let (command, continuation) = commandQueue.removeFirst()
        Task { @MainActor in
            let result = await runWithRetry(command)
            continuation.resume(returning: result)
            isProcessing = false
            processNextIfNeeded()
        }
    }

    private func runWithRetry(_ command: ObdCommand) async -> ObdCommandResult {
        for attempt in 0..<command.maxAttempts {
            switch await runSingleAttempt(command) {
            case .prompted(let raw):
                return .success(raw)
            case .timedOut:
                if attempt == command.maxAttempts - 1 { return .timeout }
            }
        }
        return .timeout
    }

    private enum AttemptOutcome {
        case prompted(String)
        case timedOut
    }

    private func runSingleAttempt(_ command: ObdCommand) async -> AttemptOutcome {
        drainStaleIncoming()
        let text = command.text.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() + "\r"
        transport.write(Data(text.utf8))

        let deadline = ContinuousClock.now.advanced(by: .milliseconds(command.timeoutMs))
        while !incomingBuffer.contains(">") {
            let remaining = deadline - ContinuousClock.now
            guard remaining > .zero else { return .timedOut }
            guard await waitForIncoming(timeout: remaining) else { return .timedOut }
        }
        let raw = incomingBuffer
        incomingBuffer = ""
        return .prompted(raw)
    }

    /// Suspends until either new data arrives (returns true) or `timeout` elapses (returns
    /// false). Both outcomes resume the same continuation, guarded so only the first one wins —
    /// the loser (usually the timeout firing after data already arrived) simply no-ops instead of
    /// leaving anything suspended, unlike a cancelled `TaskGroup` child task would.
    private func waitForIncoming(timeout: Duration) async -> Bool {
        await withCheckedContinuation { (continuation: CheckedContinuation<Bool, Never>) in
            var resumed = false
            wakeUp = {
                guard !resumed else { return }
                resumed = true
                continuation.resume(returning: true)
            }
            Task { @MainActor in
                try? await Task.sleep(for: timeout)
                guard !resumed else { return }
                resumed = true
                self.wakeUp = nil
                continuation.resume(returning: false)
            }
        }
    }
}
